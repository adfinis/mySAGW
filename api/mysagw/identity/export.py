from collections import OrderedDict


def _str_or_empty(value):
    mapping = {True: 1, False: 0, None: ""}
    return mapping.get(value, str(value))


def _fetch_related_strings(manager, field):
    return [_str_or_empty(getattr(entry, field)) for entry in manager.iterator()]


class IdentityExport:
    FIELDS = {
        "first_name": lambda i, a: i.first_name,
        "last_name": lambda i, a: i.last_name,
        "localized_salutation": lambda i, a: i.localized_salutation,
        "localized_title": lambda i, a: i.localized_title,
        "language": lambda i, a: i.language,
        "is_organisation": lambda i, a: i.is_organisation,
        "organisation_name": lambda i, a: i.organisation_name,
        "email": lambda i, a: i.email,
        "additional_emails": lambda i, a: "\n".join(
            _fetch_related_strings(i.additional_emails, "email"),
        ),
        "phone_numbers": lambda i, a: "\n".join(
            _fetch_related_strings(i.phone_numbers, "phone"),
        ),
        "address_addition_1": lambda i, a: "" if not a else a.address_addition_1,
        "address_addition_2": lambda i, a: "" if not a else a.address_addition_2,
        "address_addition_3": lambda i, a: "" if not a else a.address_addition_3,
        "street_and_number": lambda i, a: "" if not a else a.street_and_number,
        "po_box": lambda i, a: "" if not a else a.po_box,
        "postcode": lambda i, a: "" if not a else a.postcode,
        "town": lambda i, a: "" if not a else a.town,
        "country": lambda i, a: "" if not a else a.country.name,
    }

    def fetch_identity_data(self, identity, fields):
        address = None
        if identity.addresses.exists():
            address = identity.addresses.get(default=True)

        data = OrderedDict()

        for field_key, field_getter in fields.items():
            data[field_key] = _str_or_empty(field_getter(identity, address))

        return data

    def export(self, qs, include_fields=None, ignore_empty=False):
        include_fields = include_fields if include_fields else self.FIELDS.keys()
        fields = {name: self.FIELDS[name] for name in include_fields}

        records = []
        for identity in qs.iterator(2000):
            identity_data = self.fetch_identity_data(identity, fields)

            if not ignore_empty or any(identity_data.values()):
                records.append(identity_data)

        return records


class MembershipExport:
    FIELDS = {
        "first_name": lambda m, a: m.identity.first_name,
        "last_name": lambda m, a: m.identity.last_name,
        "localized_salutation": lambda m, a: m.identity.localized_salutation,
        "localized_title": lambda m, a: m.identity.localized_title,
        "language": lambda m, a: m.identity.language,
        "email": lambda m, a: m.identity.email,
        "address_addition_1": lambda m, a: "" if not a else a.address_addition_1,
        "address_addition_2": lambda m, a: "" if not a else a.address_addition_2,
        "address_addition_3": lambda m, a: "" if not a else a.address_addition_3,
        "street_and_number": lambda m, a: "" if not a else a.street_and_number,
        "po_box": lambda m, a: "" if not a else a.po_box,
        "postcode": lambda m, a: "" if not a else a.postcode,
        "town": lambda m, a: "" if not a else a.town,
        "country": lambda m, a: "" if not a else a.country.name,
        "organisation": lambda m, a: m.organisation.organisation_name,
        "role": lambda m, a: "" if not m.role else m.role.title[m.identity.language],
        "inactive": lambda m, a: m.inactive,
        "tenure": lambda m, a: ""
        if not m.time_slot
        else f"{m.time_slot.lower if m.time_slot.lower else ''} - {m.time_slot.upper if m.time_slot.upper else ''}",
        "next_election": lambda m, a: m.next_election,
    }

    def fetch_membership_data(self, membership, fields):
        address = None
        if membership.identity.addresses.exists():
            address = membership.identity.addresses.get(default=True)

        data = OrderedDict()

        for field_key, field_getter in fields.items():
            data[field_key] = _str_or_empty(field_getter(membership, address))

        return data

    def export(self, qs, include_fields=None, ignore_empty=False):
        include_fields = include_fields if include_fields else self.FIELDS.keys()
        fields = {name: self.FIELDS[name] for name in include_fields}

        records = []
        for membership in qs.iterator(2000):
            membership_data = self.fetch_membership_data(membership, fields)

            if not ignore_empty or any(membership_data.values()):
                records.append(membership_data)

        return records
