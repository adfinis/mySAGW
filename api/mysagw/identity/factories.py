from factory import Faker, Maybe, SubFactory, fuzzy
from factory.django import DjangoModelFactory

from . import models


class IdentityFactory(DjangoModelFactory):
    idp_id = Faker("uuid4")
    email = Faker("email")
    is_organisation = False
    organisation_name = Maybe("is_organisation", Faker("company"), None)
    first_name = Faker("first_name")
    last_name = Faker("last_name")
    salutation = models.Identity.SALUTATION_NEUTRAL
    title = models.Identity.TITLE_NONE
    language = "de"

    class Meta:
        model = models.Identity


class EmailFactory(DjangoModelFactory):
    identity = SubFactory(IdentityFactory)
    email = Faker("email")
    description = fuzzy.FuzzyChoice(["work", "private", "assistant"])

    class Meta:
        model = models.Email


class PhoneNumberFactory(DjangoModelFactory):
    identity = SubFactory(IdentityFactory)
    phone = Faker("phone_number")
    description = fuzzy.FuzzyChoice(["work", "private", "home", "office", "mobile"])
    default = True

    class Meta:
        model = models.PhoneNumber


class AddressFactory(DjangoModelFactory):
    identity = SubFactory(IdentityFactory)
    street_and_number = Faker("street_address")
    postcode = Faker("postcode")
    town = Faker("city")
    country = Faker("country_code")
    description = fuzzy.FuzzyChoice(["work", "private", "home", "office"])
    default = True
    address_addition_1 = None
    address_addition_2 = None
    address_addition_3 = None
    po_box = None

    class Meta:
        model = models.Address


class InterestCategoryFactory(DjangoModelFactory):
    title = Faker("word")
    description = Faker("sentence")
    archived = False
    public = False

    class Meta:
        model = models.InterestCategory


class InterestFactory(DjangoModelFactory):
    title = Faker("job")
    description = Faker("sentence")
    archived = False
    category = SubFactory(InterestCategoryFactory)

    class Meta:
        model = models.Interest


class MembershipRoleFactory(DjangoModelFactory):
    title = Faker("multilang", faker_provider="job")
    description = Faker("multilang", faker_provider="text")
    archived = False

    class Meta:
        model = models.MembershipRole


class MembershipFactory(DjangoModelFactory):
    identity = SubFactory(IdentityFactory)
    organisation = SubFactory(IdentityFactory, is_organisation=True)
    role = SubFactory(MembershipRoleFactory)
    time_slot = None
    authorized = False
    inactive = False

    class Meta:
        model = models.Membership
