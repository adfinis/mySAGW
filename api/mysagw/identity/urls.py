from django.urls import re_path
from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"additional-emails", views.EmailViewSet)
r.register(r"phone-numbers", views.PhoneNumberViewSet)
r.register(r"addresses", views.AddressViewSet)
r.register(r"identities", views.IdentityViewSet)
r.register(
    r"public-identities",
    views.PublicIdentitiesViewSet,
    basename="public-identities",
)
r.register(r"interest-categories", views.InterestCategoryViewSet)
r.register(r"interests", views.InterestViewSet)
r.register(r"membership-roles", views.MembershipRoleViewSet)
r.register(r"memberships", views.MembershipViewSet)
r.register(r"my-orgs", views.MyOrgsViewSet, basename="my-orgs")
r.register(r"my-memberships", views.MyMembershipViewSet, basename="my-memberships")
r.register(
    r"org-memberships",
    views.OrganisationAdminMembersViewSet,
    basename="org-memberships",
)

urlpatterns = [
    re_path(
        r"^me/?$",
        views.MeViewSet.as_view({"get": "retrieve", "patch": "update"}),
        name="me",
    ),
]

urlpatterns.extend(r.urls)
