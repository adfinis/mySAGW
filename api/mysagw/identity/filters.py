import operator
import shlex
from functools import reduce

from django.db.models import Max, OuterRef, Q, Subquery, Value
from django.db.models.functions import Coalesce
from django.utils import timezone
from django_filters import BooleanFilter, UUIDFilter
from django_filters.rest_framework import FilterSet
from rest_framework.compat import distinct
from rest_framework.filters import SearchFilter

from mysagw.filters import CharMultiValueFilter
from mysagw.identity import models


class IdentityFilterSet(FilterSet):
    idp_ids = CharMultiValueFilter(field_name="idp_id")
    has_idp_id = BooleanFilter(field_name="idp_id", lookup_expr="isnull", exclude=True)
    member_of_organisations = CharMultiValueFilter(
        field_name="memberships__organisation__organisation_name",
        distinct=True,
        method="member_of_organisations_filter",
    )

    @staticmethod
    def member_of_organisations_filter(queryset, name, value):
        membership_base_query = models.Membership.objects.filter(
            models.Q(organisation__organisation_name__in=value)
            & (
                (
                    models.Q(time_slot__isnull=True)
                    | models.Q(time_slot__contains=timezone.now())
                )
                & models.Q(inactive=False)
            ),
        )

        membership_base_subquery = membership_base_query.filter(
            identity_id=OuterRef("pk"),
        )
        return queryset.filter(
            memberships__id__in=Subquery(membership_base_subquery.values("id")),
        ).distinct()

    class Meta:
        model = models.Identity
        fields = [
            "email",
            "idp_id",
            "is_organisation",
            "member_of_organisations",
        ]


class EmailFilterSet(FilterSet):
    class Meta:
        model = models.Email
        fields = [
            "identity",
        ]


class PhoneNumberFilterSet(FilterSet):
    class Meta:
        model = models.PhoneNumber
        fields = [
            "identity",
        ]


class AddressFilterSet(FilterSet):
    class Meta:
        model = models.Address
        fields = [
            "identity",
        ]


class MembershipFilterSet(FilterSet):
    class Meta:
        model = models.Membership
        fields = [
            "identity",
            "organisation",
            "authorized",
        ]


class InterestFilterSet(FilterSet):
    public = BooleanFilter(field_name="category__public")

    class Meta:
        model = models.Interest
        fields = [
            "public",
        ]


class InterestCategoryFilterSet(FilterSet):
    class Meta:
        model = models.InterestCategory
        fields = [
            "public",
        ]


class OrganisationAdminMembersFilterSet(FilterSet):
    organisation = UUIDFilter(
        field_name="memberships__organisation_id",
        method="organisation_filter",
        required=True,
    )

    def organisation_filter(self, queryset, name, value):
        queryset = queryset.filter(**{name: value})
        queryset = queryset.annotate(
            highest_active_role=Coalesce(
                Max(
                    "memberships__role__sort",
                    filter=Q(**{name: value})
                    & Q(memberships__inactive=False)
                    & (
                        Q(memberships__time_slot__isnull=True)
                        | Q(memberships__time_slot__contains=timezone.now())
                    ),
                ),
                Value(0),
            ),
        )
        queryset = queryset.annotate(
            highest_inactive_role=Coalesce(
                Max(
                    "memberships__role__sort",
                    filter=Q(**{name: value})
                    & ~Q(
                        Q(memberships__inactive=False)
                        & (
                            Q(memberships__time_slot__isnull=True)
                            | Q(memberships__time_slot__contains=timezone.now())
                        ),
                    ),
                ),
                Value(0),
            ),
        )

        return queryset.order_by(
            "-highest_active_role",
            "-highest_inactive_role",
            "last_name",
            "first_name",
            "email",
        )

    class Meta:
        model = models.Identity
        fields = [
            "organisation",
        ]


class SAGWSearchFilter(SearchFilter):
    @staticmethod
    def _split(value):
        lex = shlex.shlex(value, posix=True)
        lex.whitespace_split = True
        lex.commenters = ""
        lex.quotes = '"'
        return [i.strip() for i in lex if i.strip()]

    def get_search_terms(self, request):
        params = request.query_params.get(self.search_param, "")
        params = params.replace("\x00", "")  # strip null characters
        params = params.replace(",", " ")
        try:
            return self._split(params)
        except ValueError as e:
            if e.args[0] == "No closing quotation":
                return self._split(f'{params}"')

    def filter_queryset(self, request, queryset, view):  # noqa: C901
        search_fields = self.get_search_fields(view, request)
        search_terms = self.get_search_terms(request)

        if not search_fields or not search_terms:
            return queryset

        orm_lookups = [
            self.construct_search(str(search_field)) for search_field in search_fields
        ]

        base = queryset
        # we're only interested in active memberships
        membership_base_query = models.Membership.objects.filter(
            (
                (
                    models.Q(time_slot__isnull=True)
                    | models.Q(time_slot__contains=timezone.now())
                )
                & models.Q(inactive=False)
            ),
        )

        membership_base_subquery = membership_base_query.filter(
            identity_id=OuterRef("pk"),
        )

        filters = []
        excludes = []

        for search_term in search_terms:
            exclude = False
            if search_term.startswith("-"):
                exclude = True
                search_term = search_term.lstrip("-")

            queries = []
            for orm_lookup in orm_lookups:
                lookup = models.Q(**{orm_lookup: search_term})

                if orm_lookup.startswith("memberships__"):
                    membership_orm_lookup = orm_lookup.replace("memberships__", "")
                    if not membership_base_query.filter(
                        **{membership_orm_lookup: search_term},
                    ).exists():
                        # short-circuiting in order to save ~75% of processing time
                        continue

                    membership_subquery = membership_base_subquery.filter(
                        **{membership_orm_lookup: search_term},
                    )
                    # first Q object short circuits again, for a performance gain
                    lookup = models.Q(**{orm_lookup: search_term}) & models.Q(
                        memberships__id__in=Subquery(membership_subquery.values("id")),
                    )

                queries.append(lookup)

            condition = reduce(operator.or_, queries)
            if exclude:
                excludes.append(condition)
            else:
                filters.append(condition)

        needed = [set(queryset.filter(f).values_list("pk", flat=True)) for f in filters]

        if needed == [set()]:
            # if no records match our filters, we don't care about excludes
            return models.Identity.objects.none()

        excluded = [
            pk
            for f in excludes
            for pk in queryset.filter(f).values_list("pk", flat=True)
        ]

        if not needed:
            # only excludes have been provided to the filter
            queryset = queryset.exclude(pk__in=excluded)
        else:
            needed_pks = reduce(lambda a, b: a.intersection(b), needed)
            queryset = queryset.filter(pk__in=needed_pks).exclude(pk__in=excluded)

        if self.must_call_distinct(queryset, search_fields):
            # Filtering against a many-to-many field requires us to
            # call queryset.distinct() in order to avoid duplicate items
            # in the resulting queryset.
            # We try to avoid this if possible, for performance reasons.
            queryset = distinct(queryset, base)

        return queryset
