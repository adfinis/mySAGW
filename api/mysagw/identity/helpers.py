from mysagw.identity.models import MembershipRole


def get_membership_roles_order():
    roles_order = [
        [r.sort, r.title.de, str(r.pk)] for r in MembershipRole.objects.iterator()
    ]
    print(roles_order)
    return roles_order


def set_membership_roles_order(roles_order):
    """
    Set the order of the roles.

    This function accepts a list in the same format of the output of
    `get_roles_order()`.
    """
    for raw_role in roles_order:
        role = MembershipRole.objects.get(pk=raw_role[2])
        role.sort = raw_role[0]
        role.save()
