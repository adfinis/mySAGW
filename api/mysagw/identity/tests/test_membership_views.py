import datetime

import pytest
from django.conf import settings
from django.urls import reverse
from psycopg2.extras import DateRange
from rest_framework import status
from syrupy import filters

from mysagw.identity import models


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_200_OK),
        ("staff", status.HTTP_200_OK),
        ("admin", status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_role_detail(db, membership_role, client, expected_status):
    url = reverse("membershiprole-detail", args=[membership_role.pk])

    response = client.get(url)

    assert response.status_code == expected_status

    json = response.json()
    assert json["data"]["id"] == str(membership_role.pk)


@pytest.mark.parametrize(
    "client,own,expected_status",
    [
        ("user", False, status.HTTP_403_FORBIDDEN),
        ("user", True, status.HTTP_403_FORBIDDEN),
        ("staff", False, status.HTTP_200_OK),
        ("admin", False, status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_detail(db, membership, client, own, expected_status):
    if own:
        membership.identity = client.user.identity
        membership.save()

    url = reverse("membership-detail", args=[membership.pk])

    response = client.get(url, {"include": "role"})

    assert response.status_code == expected_status

    if expected_status != status.HTTP_200_OK:
        return

    json = response.json()
    assert json["data"]["id"] == str(membership.pk)
    assert json["included"][0]["id"] == str(membership.role.pk)


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_200_OK),
        ("staff", status.HTTP_200_OK),
        ("admin", status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_role_list(db, client, expected_status, membership_role_factory):
    membership_role_factory.create_batch(3)

    url = reverse("membershiprole-list")

    response = client.get(url)

    assert response.status_code == expected_status

    json = response.json()
    assert len(json["data"]) == models.MembershipRole.objects.count() == 3


@pytest.mark.parametrize("lang", ["de", "fr"])
def test_membership_role_list_ordering(db, client, lang, membership_role_factory):
    role1 = membership_role_factory(title={"de": "aaa", "fr": "bbb"})
    role2 = membership_role_factory(title={"de": "bbb", "fr": "aaa"})

    expected = [str(role1.pk), str(role2.pk)]
    if lang == "fr":
        expected = [str(role2.pk), str(role1.pk)]

    url = reverse("membershiprole-list")

    response = client.get(url, HTTP_ACCEPT_LANGUAGE=lang)

    assert response.status_code == status.HTTP_200_OK

    json = response.json()

    assert len(json["data"]) == 2
    assert json["data"][0]["id"] == expected[0]
    assert json["data"][1]["id"] == expected[1]


@pytest.mark.parametrize(
    "client,expected_count,expected_status",
    [
        ("user", 0, status.HTTP_403_FORBIDDEN),
        ("staff", 3, status.HTTP_200_OK),
        ("admin", 3, status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_list(
    db, client, expected_count, expected_status, membership_factory
):
    membership_factory.create_batch(2)
    membership_factory(identity=client.user.identity)

    url = reverse("membership-list")

    response = client.get(url)

    assert response.status_code == expected_status

    if expected_status == status.HTTP_200_OK:
        json = response.json()
        assert len(json["data"]) == expected_count


@pytest.mark.parametrize(
    "client,expected_count",
    [
        ("user", 1),
        ("staff", 1),
        ("admin", 1),
    ],
    indirect=["client"],
)
def test_my_membership_list(db, client, expected_count, membership_factory):
    membership_factory.create_batch(2)
    membership_factory(identity=client.user.identity)

    url = reverse("my-memberships-list")

    response = client.get(url)

    assert response.status_code == status.HTTP_200_OK

    json = response.json()
    assert len(json["data"]) == expected_count


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_201_CREATED),
        ("admin", status.HTTP_201_CREATED),
    ],
    indirect=["client"],
)
def test_membership_role_create(db, client, expected_status):
    url = reverse("membershiprole-list")

    data = {
        "data": {"type": "membership-roles", "attributes": {"title": {"de": "Foo"}}}
    }

    response = client.post(url, data=data)

    assert response.status_code == expected_status

    if expected_status == status.HTTP_403_FORBIDDEN:
        return

    assert models.MembershipRole.objects.count() == 1
    assert dict(models.MembershipRole.objects.first().title) == {
        "de": "Foo",
        "en": "",
        "fr": "",
    }


@pytest.mark.parametrize(
    "title",
    [
        {"de": ""},
        {"de": None},
        {"en": "Foo"},
        {},
    ],
)
def test_membership_role_empty_failure(db, client, title):
    url = reverse("membershiprole-list")
    data = {"data": {"type": "membership-roles", "attributes": {"title": title}}}

    response = client.post(url, data=data)

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert (
        response.json()["errors"][0]["detail"]
        == f'Title must be set for language: "{settings.LANGUAGE_CODE}"'
    )


@pytest.mark.parametrize(
    "client,identity_is_organisation,organisation_is_organisation,set_lower,set_upper,expected_status",
    [
        ("user", False, True, True, True, status.HTTP_403_FORBIDDEN),
        ("staff", False, True, True, True, status.HTTP_201_CREATED),
        ("admin", False, True, True, True, status.HTTP_201_CREATED),
        ("admin", False, True, False, True, status.HTTP_201_CREATED),
        ("admin", False, True, True, False, status.HTTP_201_CREATED),
        ("admin", False, True, False, False, status.HTTP_201_CREATED),
        ("admin", True, True, True, True, status.HTTP_400_BAD_REQUEST),
        ("admin", False, False, True, True, status.HTTP_400_BAD_REQUEST),
    ],
    indirect=["client"],
)
def test_membership_create(
    db,
    client,
    set_lower,
    set_upper,
    expected_status,
    identity_factory,
    identity_is_organisation,
    organisation_is_organisation,
):
    identity = identity_factory(is_organisation=identity_is_organisation)
    assert identity.modified_by_user != client.user.id
    organisation = identity_factory(is_organisation=organisation_is_organisation)

    url = reverse("membership-list")

    start = datetime.date(2020, 1, 1) if set_lower else None
    end = datetime.date(2020, 12, 31) if set_upper else None
    bounds = "[)" if set_lower else "()"
    next_election = datetime.date(2020, 11, 23)

    data = {
        "data": {
            "type": "memberships",
            "attributes": {
                "time-slot": {"lower": start, "upper": end},
                "next-election": next_election,
            },
            "relationships": {
                "identity": {"data": {"id": str(identity.pk), "type": "identities"}},
                "organisation": {
                    "data": {"id": str(organisation.pk), "type": "identities"}
                },
            },
        }
    }

    response = client.post(url, data=data)

    assert response.status_code == expected_status

    if expected_status != status.HTTP_201_CREATED:
        return

    assert models.Membership.objects.count() == 1
    membership = models.Membership.objects.first()
    assert membership.identity == identity
    assert membership.organisation == organisation
    assert membership.time_slot == DateRange(start, end, bounds)
    assert membership.next_election == next_election
    assert membership.identity.modified_by_user == client.user.id


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_403_FORBIDDEN),
        ("admin", status.HTTP_403_FORBIDDEN),
    ],
    indirect=["client"],
)
def test_my_membership_create(
    db,
    client,
    expected_status,
    identity_factory,
):
    identity = identity_factory(is_organisation=False)
    assert identity.modified_by_user != client.user.id
    organisation = identity_factory(is_organisation=True)

    url = reverse("my-memberships-list")

    data = {
        "data": {
            "type": "memberships",
            "attributes": {
                "time-slot": {"lower": None, "upper": None},
                "next-election": None,
            },
            "relationships": {
                "identity": {"data": {"id": str(identity.pk), "type": "identities"}},
                "organisation": {
                    "data": {"id": str(organisation.pk), "type": "identities"}
                },
            },
        }
    }

    response = client.post(url, data=data)

    assert response.status_code == expected_status


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_200_OK),
        ("admin", status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_role_update(db, client, expected_status, membership_role_factory):
    membershiprole = membership_role_factory(title="Bar")
    url = reverse("membershiprole-detail", args=[membershiprole.pk])

    data = {
        "data": {
            "type": "membership-roles",
            "id": str(membershiprole.pk),
            "attributes": {"title": {"de": "Foo"}},
        }
    }

    response = client.patch(url, data=data)

    assert response.status_code == expected_status

    if expected_status != status.HTTP_200_OK:
        return

    json = response.json()
    assert json["data"]["attributes"]["title"] == {"de": "Foo", "en": "", "fr": ""}
    membershiprole.refresh_from_db()
    assert dict(membershiprole.title) == {"de": "Foo", "en": "", "fr": ""}


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_200_OK),
        ("admin", status.HTTP_200_OK),
    ],
    indirect=["client"],
)
def test_membership_update(db, client, expected_status, membership_factory):
    membership = membership_factory()

    url = reverse("membership-detail", args=[membership.pk])

    data = {
        "data": {
            "type": "memberships",
            "id": str(membership.pk),
            "attributes": {"comment": "Foo"},
            "relationships": {"role": {}},
        },
    }

    response = client.patch(url, data=data)

    assert response.status_code == expected_status

    if expected_status != status.HTTP_200_OK:
        return

    membership.refresh_from_db()
    assert membership.role is None
    assert membership.comment == "Foo"


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_403_FORBIDDEN),
        ("admin", status.HTTP_403_FORBIDDEN),
    ],
    indirect=["client"],
)
def test_my_membership_update(db, client, expected_status, membership_factory):
    membership = membership_factory()

    url = reverse("my-memberships-detail", args=[membership.pk])

    data = {
        "data": {
            "type": "memberships",
            "id": str(membership.pk),
            "attributes": {"comment": "Foo"},
            "relationships": {"role": {}},
        },
    }

    response = client.patch(url, data=data)

    assert response.status_code == expected_status


@pytest.mark.parametrize("field", ["identity", "organisation"])
def test_membership_update_failure(db, client, field, membership, identity_factory):
    identity = identity_factory(is_organisation=field == "organisation")

    url = reverse("membership-detail", args=[membership.pk])

    data = {
        "data": {
            "type": "memberships",
            "id": str(membership.pk),
            "relationships": {
                field: {"data": {"type": "identities", "id": str(identity.pk)}},
            },
        }
    }

    response = client.patch(url, data=data)

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    assert (
        response.json()["errors"][0]["detail"] == f'Field "{field}" can\'t be modified.'
    )


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_204_NO_CONTENT),
        ("admin", status.HTTP_204_NO_CONTENT),
    ],
    indirect=["client"],
)
def test_membership_role_delete(db, client, expected_status, membership_role_factory):
    membershiprole = membership_role_factory()
    url = reverse("membershiprole-detail", args=[membershiprole.pk])

    response = client.delete(url)

    assert response.status_code == expected_status

    if expected_status == status.HTTP_204_NO_CONTENT:
        with pytest.raises(models.MembershipRole.DoesNotExist):
            membershiprole.refresh_from_db()


@pytest.mark.parametrize(
    "client,expected_status",
    [
        ("user", status.HTTP_403_FORBIDDEN),
        ("staff", status.HTTP_204_NO_CONTENT),
        ("admin", status.HTTP_204_NO_CONTENT),
    ],
    indirect=["client"],
)
def test_membership_delete(db, client, expected_status, membership_factory):
    membership = membership_factory()
    assert membership.identity.modified_by_user != client.user.id
    url = reverse("membership-detail", args=[membership.pk])

    response = client.delete(url)

    assert response.status_code == expected_status

    if expected_status == status.HTTP_204_NO_CONTENT:
        with pytest.raises(models.Membership.DoesNotExist):
            membership.refresh_from_db()
        membership.identity.refresh_from_db()
        assert membership.identity.modified_by_user == client.user.id


@pytest.mark.parametrize(
    "client",
    ["user"],
    indirect=["client"],
)
def test_my_memberships_delete_failure(db, client):
    url = reverse("my-memberships-detail", args=["foo"])

    response = client.delete(url)

    assert response.status_code == status.HTTP_403_FORBIDDEN


def test_organisation_members(
    db, client, membership_role_factory, membership_factory, identity_factory, snapshot
):
    first_role = membership_role_factory.create(sort=1)
    second_role = membership_role_factory.create(sort=2)
    third_role = membership_role_factory.create(sort=3)
    organisation = identity_factory(is_organisation=True)
    user1 = identity_factory(first_name="User", last_name="1")
    user2 = identity_factory(first_name="User", last_name="2")
    user3 = identity_factory(first_name="User", last_name="3")
    user4 = identity_factory(first_name="User", last_name="4")
    user5 = identity_factory(first_name="User", last_name="5")

    # User1, 2 memberships, 1 inactive
    # second role is highest
    # should be second in list - same highest role as user 3, but last_name comes first
    # in alphabet
    membership_factory.create(
        role=first_role, identity=user1, organisation=organisation
    )
    membership_factory.create(
        role=second_role, identity=user1, organisation=organisation
    )
    membership_factory.create(
        role=third_role,
        identity=user1,
        organisation=organisation,
        time_slot=DateRange(datetime.date(2020, 1, 1), datetime.date(2020, 12, 31)),
    )

    # User2, 3 memberships, 1 inactive
    # third role is highest
    # should be first in list
    membership_factory.create(
        role=second_role, identity=user2, organisation=organisation, inactive=True
    )
    membership_factory.create(
        role=second_role, identity=user2, organisation=organisation
    )
    membership_factory.create(
        role=third_role, identity=user2, organisation=organisation
    )

    # User3, 2 memberships, 1 inactive
    # second role highest
    # should be third in list - same highest role as user 1, but last_name comes after
    # in alphabet
    membership_factory.create(
        role=second_role, identity=user3, organisation=organisation
    )
    membership_factory.create(
        role=third_role, identity=user3, organisation=organisation, inactive=True
    )

    # User4, 1 membership, 1 inactive
    # should be fifth in list
    membership_factory.create(
        role=second_role, identity=user4, organisation=organisation, inactive=True
    )

    # User5, 1 membership, 1 inactive
    # should be fourth in list
    membership_factory.create(
        role=third_role, identity=user5, organisation=organisation, inactive=True
    )

    url = reverse("org-memberships-list")

    response = client.get(url, {"filter[organisation]": str(organisation.pk)})

    assert response.status_code == status.HTTP_200_OK

    json = response.json()
    assert len(json["data"]) == 5
    assert json["data"][0]["attributes"]["last-name"] == "2"
    assert json["data"][1]["attributes"]["last-name"] == "1"
    assert json["data"][2]["attributes"]["last-name"] == "3"
    assert json["data"][3]["attributes"]["last-name"] == "5"
    assert json["data"][4]["attributes"]["last-name"] == "4"
    assert json == snapshot(exclude=filters.props("id"))
