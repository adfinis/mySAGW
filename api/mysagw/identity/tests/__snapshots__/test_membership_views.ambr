# name: test_organisation_members[admin]
  dict({
    'data': list([
      dict({
        'attributes': dict({
          'email': 'evansthomas@example.com',
          'first-name': 'User',
          'last-name': '2',
          'roles': list([
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Beamter',
                'en': 'Tour manager',
                'fr': 'officier de la marine marchande',
              }),
            }),
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Altenpfleger',
                'en': 'Investment banker, corporate',
                'fr': 'ingénieur efficacité énergétique du bâtiment',
              }),
            }),
            dict({
              'authorized': False,
              'inactive': True,
              'role': dict({
                'de': 'Altenpfleger',
                'en': 'Investment banker, corporate',
                'fr': 'ingénieur efficacité énergétique du bâtiment',
              }),
            }),
          ]),
        }),
        'type': 'identities',
      }),
      dict({
        'attributes': dict({
          'email': 'michelle86@example.com',
          'first-name': 'User',
          'last-name': '1',
          'roles': list([
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Altenpfleger',
                'en': 'Investment banker, corporate',
                'fr': 'ingénieur efficacité énergétique du bâtiment',
              }),
            }),
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Drogist',
                'en': 'Educational psychologist',
                'fr': 'journaliste',
              }),
            }),
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Beamter',
                'en': 'Tour manager',
                'fr': 'officier de la marine marchande',
              }),
            }),
          ]),
        }),
        'type': 'identities',
      }),
      dict({
        'attributes': dict({
          'email': 'ebonyjohnson@example.com',
          'first-name': 'User',
          'last-name': '3',
          'roles': list([
            dict({
              'authorized': False,
              'inactive': False,
              'role': dict({
                'de': 'Altenpfleger',
                'en': 'Investment banker, corporate',
                'fr': 'ingénieur efficacité énergétique du bâtiment',
              }),
            }),
            dict({
              'authorized': False,
              'inactive': True,
              'role': dict({
                'de': 'Beamter',
                'en': 'Tour manager',
                'fr': 'officier de la marine marchande',
              }),
            }),
          ]),
        }),
        'type': 'identities',
      }),
      dict({
        'attributes': dict({
          'email': 'johnsonmercedes@example.net',
          'first-name': 'User',
          'last-name': '5',
          'roles': list([
            dict({
              'authorized': False,
              'inactive': True,
              'role': dict({
                'de': 'Beamter',
                'en': 'Tour manager',
                'fr': 'officier de la marine marchande',
              }),
            }),
          ]),
        }),
        'type': 'identities',
      }),
      dict({
        'attributes': dict({
          'email': 'delacruzbrandon@example.com',
          'first-name': 'User',
          'last-name': '4',
          'roles': list([
            dict({
              'authorized': False,
              'inactive': True,
              'role': dict({
                'de': 'Altenpfleger',
                'en': 'Investment banker, corporate',
                'fr': 'ingénieur efficacité énergétique du bâtiment',
              }),
            }),
          ]),
        }),
        'type': 'identities',
      }),
    ]),
  })
# ---
